### ISSUE TITLE
xcode_proj handling of bazel artifacts

### ISSUE THREAD

**Issue by jerrymarino** [bazel-ios/rules_ios#190]
I've noticed a the `xcode_project` rule is pointing at `bazel-out` for a few artifacts. For example `HEADER_SEARCH_PATHS`. This might cause some issues in the case where the files don't exist: cleaning in Xcode, opening Xcode prior to building. Additionally it will cause issues if those files are outputs of a simulator build and the user is never using a simulator. In that case, the indexer and autocomplete will not work.

For what it's worth, XCHammer does a few things to get around this - manage generated files relevant to indexing and building on it's own and in some cases building them up front.

**Comment by jerrymarino** [comment]
This commit shows an example of _how_ the bazel configuration output dir hash is factored in to some project settings:
https://github.com/bazel-ios/rules_ios/pull/188/commits/02f3c362eaf34623923e51393557c39790903202

**Comment by gyfelton** [comment]
Thanks for identifying this problem! Just to confirm it is the method `_header_search_paths_for_target` in question right?
What you said is consistent with the observation that indexing won't work unless certain targets are built. Am I correct on this?

**Comment by thiagohmcruz** [comment]
@jerrymarino mind sharing one example from XCHammer code where this kind of handling is happening? Experiencing an issue now where "jump to definition" takes me to the file in the `bazel-out` folder instead of the one inside the project generated by the `xcodeproj` rule.

>I've noticed a the xcode_project rule is pointing at bazel-out for a few artifacts
>

Would appreciate if you could share exactly how you concluded this, what file did you inspect/command did you run?

**Comment by jerrymarino** [comment]
@gyfelton yeah one of the issues is that the header search paths are pointing at simulator intermediates so if you were just doing device builds those intermediates wouldn't exist.

@thiagohmcruz XCHammer today works similar to how cocoapods generator does: it puts all the targets into Xcode project with equivalent bazel build settings. Because of this, the features in the UI generally work as they did with cocoapods. The main caveat is that if you flip the bit to build with Bazel it calls Tulsi's infrastructure to run bazel.

To achieve this there are [tens of rule attributes and features](https://github.com/pinterest/xchammer/blob/master/Sources/XCHammer/XcodeTarget.swift). Today, rules_ios handles files, framework search paths, and header search paths.

If we wanted to try XCHammer's approach, we could try to leverage the `xcode_project` rule similar to how `ios_application` call's rules_ios `ios_application`  and pass any `rules_ios` specific flags. As `rules_ios` uses `xcodegen` as well, I suspect it should mesh w/o significant changes

```
# xcodeproj.bzl

load("@xchammer//:xcodeproject.bzl", xcodegen_bazel_xcodeproj="xcode_project")
def xcodegen_bazel_xcodeproj(kwargs):
  xcode_project(
       name = ...,
       targets = ...
       paths = [ "**" ],
  )
```

**Comment by justinseanmartin** [comment]
@amberdixon ^ I believe this was an explicit decision in the way that Amber set up the project generator.

**Comment by amberdixon** [comment]
Yes, the expected behavior is that if you have performed a build locally, the header search paths and framework search paths should point to the files generated by the build under the bazel-out directory.

IIRC regular xcodeprojects generated by cocoapods behave in a similar way. If you haven't build a target, then the files for that target will not always index successfully and so autocomplete operations will initially not work.

> For what it's worth, XCHammer does a few things to get around this - manage generated files relevant to indexing and building on it's own and in some cases building them up front.

We can certainly make change to kick off some background build jobs.

There is a separate (but similar) problem that Jerry mentioned above where the device and simulator builds generate headers and framework search paths with different configurations, and the HEADER_SEARCH_PATHS/FRAMEWORK_SEARCH_PATHS build setting only points to one of those configurations. Therefore, if you build for device, but then point Xcode to build for simulator, autocomplete might fail for source files. There is probably a simple fix here, which is to load both sets of header and framework search paths into the project.

**Comment by jerrymarino** [comment]
Ah makes sense to me. Yeah theres' a few manifestations of this implementation - the one that really sticks out is indexing and code complete issues when doing device builds
